From aaa00f16a64c1c60d15328ade14c399b877738a4 Mon Sep 17 00:00:00 2001
From: John Johansen <john.johansen@canonical.com>
Date: Fri, 18 Mar 2016 06:05:39 -0700
Subject: [PATCH 1612/1767] Revert "UBUNTU: SAUCE: apparmor: ensure that
 repacedby sharing is done correctly"

BugLink: http://bugs.launchpad.net/bugs/1379535

This reverts commit b91470b3c68da4e10aeacfa4dac142327123093f.

Signed-off-by: Tim Gardner <tim.gardner@canonical.com>
---
 security/apparmor/label.c | 19 +++++++------------
 1 file changed, 7 insertions(+), 12 deletions(-)

diff --git a/security/apparmor/label.c b/security/apparmor/label.c
index e0681e4..02742b2 100644
--- a/security/apparmor/label.c
+++ b/security/apparmor/label.c
@@ -87,14 +87,6 @@ void __aa_update_replacedby(struct aa_label *orig, struct aa_label *new)
 	aa_put_label(tmp);
 }
 
-void __share_replacedby(struct aa_label *old, struct aa_label *new)
-{
-	struct aa_replacedby *r = new->replacedby;
-	new->replacedby = aa_get_replacedby(old->replacedby);
-	__aa_update_replacedby(old, new);
-	aa_put_replacedby(r);
-}
-
 /* helper fn for label_for_each_confined */
 int aa_label_next_confined(struct aa_label *l, int i)
 {
@@ -448,9 +440,10 @@ bool aa_label_replace(struct aa_label *old, struct aa_label *new)
 	if (old->hname == new->hname && labels_ns(old) == labels_ns(new)) {
 		write_lock_irqsave(&labels_set(old)->lock, flags);
 		if (old->replacedby != new->replacedby) {
-			__share_replacedby(old, new);
-		} else
-			__aa_update_replacedby(old, new);
+			free_replacedby(new->replacedby);
+			new->replacedby = aa_get_replacedby(old->replacedby);
+		}
+		__aa_update_replacedby(old, new);
 		res = __aa_label_replace(labels_set(old), old, new);
 		write_unlock_irqrestore(&labels_set(old)->lock, flags);
 	} else {
@@ -711,7 +704,9 @@ static struct aa_label *__aa_label_insert(struct aa_labelset *ls,
 				/* queued for destruction, in place replace */
 			} else {
 				if (this->replacedby) {
-					__share_replacedby(this, l);
+					free_replacedby(l->replacedby);
+					l->replacedby = aa_get_replacedby(this->replacedby);
+					__aa_update_replacedby(this, l);
 				} else
 					this->replacedby = aa_get_replacedby(l->replacedby);
 			}
-- 
1.9.3

